In [2]:
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import math
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
import pandas as pd
from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv('Speed.csv', encoding ='unicode_escape')
df.head(8)

iid   id  gender  idg  condtn  wave  round  position  positin1  order  ...  \
0    1  1.0       0    1       1     1     10         7       NaN      4  ...   
1    1  1.0       0    1       1     1     10         7       NaN      3  ...   
2    1  1.0       0    1       1     1     10         7       NaN     10  ...   
3    1  1.0       0    1       1     1     10         7       NaN      5  ...   
4    1  1.0       0    1       1     1     10         7       NaN      7  ...   
5    1  1.0       0    1       1     1     10         7       NaN      6  ...   
6    1  1.0       0    1       1     1     10         7       NaN      1  ...   
7    1  1.0       0    1       1     1     10         7       NaN      2  ...   

   attr3_3  sinc3_3  intel3_3  fun3_3  amb3_3  attr5_3  sinc5_3  intel5_3  \
0      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
1      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
2      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
3      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
4      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
5      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
6      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   
7      5.0      7.0       7.0     7.0     7.0      NaN      NaN       NaN   

   fun5_3  amb5_3  
0     NaN     NaN  
1     NaN     NaN  
2     NaN     NaN  
3     NaN     NaN  
4     NaN     NaN  
5     NaN     NaN  
6     NaN     NaN  
7     NaN     NaN  

[8 rows x 195 columns]

In [9]:
print(df.columns) 

Index(['iid', 'id', 'gender', 'idg', 'condtn', 'wave', 'round', 'position',
       'positin1', 'order',
       ...
       'attr3_3', 'sinc3_3', 'intel3_3', 'fun3_3', 'amb3_3', 'attr5_3',
       'sinc5_3', 'intel5_3', 'fun5_3', 'amb5_3'],
      dtype='object', length=195)


In [12]:
df.corr()

iid        id    gender       idg    condtn      wave  \
iid       1.000000  0.117109  0.051937  0.120362  0.220413  0.996713   
id        0.117109  1.000000 -0.011623  0.998029  0.318768  0.091571   
gender    0.051937 -0.011623  1.000000  0.033645 -0.002627 -0.001264   
idg       0.120362  0.998029  0.033645  1.000000  0.324838  0.092645   
condtn    0.220413  0.318768 -0.002627  0.324838  1.000000  0.219627   
...            ...       ...       ...       ...       ...       ...   
attr5_3   0.092325 -0.018183 -0.103692 -0.020785  0.068552  0.092906   
sinc5_3  -0.053698 -0.007993 -0.289694 -0.022967 -0.063407 -0.028111   
intel5_3  0.006949 -0.088398  0.076538 -0.086071  0.068622  0.000644   
fun5_3    0.100164 -0.046558 -0.067432 -0.050316  0.114254  0.103025   
amb5_3    0.055229 -0.119108  0.055267 -0.118940  0.145598  0.057933   

             round  position  positin1     order  ...   attr3_3   sinc3_3  \
iid       0.231881  0.081287  0.070443  0.090696  ... -0.003738 -0.005157   
id        0.374462  0.168382  0.174008  0.149216  ...  0.024213 -0.041815   
gender    0.020630 -0.001091  0.005853  0.006685  ... -0.125433 -0.159709   
idg       0.387715  0.171495  0.177639  0.154691  ...  0.020597 -0.051671   
condtn    0.816316  0.329120  0.308641  0.324108  ...  0.082965  0.034376   
...            ...       ...       ...       ...  ...       ...       ...   
attr5_3   0.025047  0.073360  0.201356  0.009776  ...  0.851922  0.177047   
sinc5_3  -0.027437  0.099616  0.102406 -0.010260  ...  0.222731  0.614853   
intel5_3  0.000923  0.093856  0.112979 -0.000037  ...  0.351189  0.275642   
fun5_3    0.105866  0.148984  0.236393  0.043915  ...  0.388431  0.172901   
amb5_3    0.110299  0.075472  0.052459  0.044849  ...  0.214301  0.146923   

          intel3_3    fun3_3    amb3_3   attr5_3   sinc5_3  intel5_3  \
iid      -0.117348  0.039944 -0.020409  0.092325 -0.053698  0.006949   
id       -0.051621 -0.029917  0.010202 -0.018183 -0.007993 -0.088398   
gender    0.023925 -0.144642 -0.058300 -0.103692 -0.289694  0.076538   
idg      -0.051081 -0.036368  0.007404 -0.020785 -0.022967 -0.086071   
condtn    0.067836  0.066229  0.023614  0.068552 -0.063407  0.068622   
...            ...       ...       ...       ...       ...       ...   
attr5_3   0.444624  0.400945  0.222258  1.000000  0.197948  0.378176   
sinc5_3   0.313755  0.184837  0.130029  0.197948  1.000000  0.505440   
intel5_3  0.681735  0.251750  0.263402  0.378176  0.505440  1.000000   
fun5_3    0.292684  0.775198  0.336507  0.406725  0.331171  0.270181   
amb5_3    0.349938  0.442050  0.610380  0.205368  0.268553  0.424972   

            fun5_3    amb5_3  
iid       0.100164  0.055229  
id       -0.046558 -0.119108  
gender   -0.067432  0.055267  
idg      -0.050316 -0.118940  
condtn    0.114254  0.145598  
...            ...       ...  
attr5_3   0.406725  0.205368  
sinc5_3   0.331171  0.268553  
intel5_3  0.270181  0.424972  
fun5_3    1.000000  0.398011  
amb5_3    0.398011  1.000000  

[187 rows x 187 columns]

### make subset out of the varibles "dec_o","age", "idg","round","match","attr_o","like_o","prob_o". I Think age is imported when someone is one a date and the attreation, like are probobly really imported to. I dont think the other varibles are as importend. 

In [25]:
df_subset = df[["dec_o","age", "idg","round","match","attr_o","like_o","prob_o"]]
df_subset = df_subset.dropna()
df_subset


dec_o   age  idg  round  match  attr_o  like_o  prob_o
0         0  21.0    1     10      0     6.0     7.0     4.0
1         0  21.0    1     10      0     7.0     8.0     4.0
2         1  21.0    1     10      1    10.0    10.0    10.0
3         1  21.0    1     10      1     7.0     7.0     7.0
4         1  21.0    1     10      1     8.0     8.0     6.0
...     ...   ...  ...    ...    ...     ...     ...     ...
8373      1  25.0   44     22      0    10.0     6.0     1.0
8374      0  25.0   44     22      0     6.0     2.0     2.0
8375      0  25.0   44     22      0     2.0     2.0     1.0
8376      1  25.0   44     22      0     5.0     6.0     4.0
8377      1  25.0   44     22      0     8.0     8.0     5.0

[7949 rows x 8 columns]

### The head()of the resulting data frame


In [18]:
X = df_subset[["age", "idg","round","match","attr_o","like_o","prob_o"]] #create the X matrix
#X = normalize(X) #normalize the matrix to put everything on the same scale
y = df_subset['dec_o'] #create the y-variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables

In [19]:
df_subset.corr()

dec_o       age       idg     round     match    attr_o    like_o  \
dec_o   1.000000 -0.049066  0.004067 -0.028637  0.519934  0.490858  0.513843   
age    -0.049066  1.000000  0.035426  0.003586 -0.035692 -0.051338 -0.040317   
idg     0.004067  0.035426  1.000000  0.391202  0.000676  0.026883 -0.000018   
round  -0.028637  0.003586  0.391202  1.000000 -0.028455 -0.050328 -0.067867   
match   0.519934 -0.035692  0.000676 -0.028455  1.000000  0.262011  0.304931   
attr_o  0.490858 -0.051338  0.026883 -0.050328  0.262011  1.000000  0.667704   
like_o  0.513843 -0.040317 -0.000018 -0.067867  0.304931  0.667704  1.000000   
prob_o  0.308246 -0.009632 -0.004731 -0.071429  0.254778  0.262328  0.463039   

          prob_o  
dec_o   0.308246  
age    -0.009632  
idg    -0.004731  
round  -0.071429  
match   0.254778  
attr_o  0.262328  
like_o  0.463039  
prob_o  1.000000

# Choose one of the following: k-nearest neighbor, decision tree or random forest
### i Choose for the k-nearest neighbor
###### The k-nearest neighbor is based on the idea that the class is equal to the ones that are close by. So if K = 10 it will check for the closes 9 'neighbors'.It check if the closes neighbors have the same features



In [20]:
X = df_subset[["age", "idg","round","match","attr_o","like_o","prob_o"]] #create the X matrix
y = df_subset['dec_o'] #create the y-variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables

# Evaluate the predictive performance of your model on the test set


In [26]:
from sklearn.neighbors import KNeighborsClassifier #the object class we need

knn = KNeighborsClassifier(n_neighbors=4) #create a KNN-classifier with 3 neighbors (default)
knn = knn.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data
knn.score(X_test, y_test) #calculate the fit on the test data
# the score is the accurracy

0.7354297693920335

### This shows that the accurracy of this model is 73,5%
###### now we make a confusion matrix to see how mutch the model is off


In [27]:
from sklearn.metrics import confusion_matrix
y_test_pred = knn.predict(X_test) #the predicted values
con = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix"
con

array([[1176,  197],
       [ 434,  578]])

In [28]:
conf_matrix = pd.DataFrame(con, index=['actual yes to a follow up date', 'actual no to follow up date'], columns = ['predicted yes', 'predicted no']) 
conf_matrix

predicted yes  predicted no
actual yes to a follow up date           1176           197
actual no to follow up date               434           578

### Accuracy: Overall, how often is the classifier correct?
(TP+TN)/total = (1176+578)/2385 = 0,73

### Misclassification Rate: Overall, how often is it wrong?
(FP+FN)/total = (434+197)/2385 = 0.26

### True Positive Rate (recall): When it's actually yes, how often does it predict yes?
TP/actual yes = 1175/1372 = 0.86%

### False Positive Rate: When it's actually no, how often does it predict yes?
FP/actual no = 434/1012 = 0,42

### Precision: When it predicts yes, how often is it correct?
TP/predicted yes = 1176/1610 = 0.73